homeassistant/components/weatherflow/config_flow.py

```diff
@@ -0,0 +1,74 @@
+"""Config flow for WeatherFlow."""
+from __future__ import annotations
+
+import asyncio
+from asyncio import Future
+from asyncio.exceptions import CancelledError
+from typing import Any
+
```

Comment:
```diff
diff --git a/homeassistant/components/weatherflow/config_flow.py b/homeassistant/components/weatherflow/config_flow.py
index 12d94dd212f..c1f24150bcc 100644
--- a/homeassistant/components/weatherflow/config_flow.py
+++ b/homeassistant/components/weatherflow/config_flow.py
@@ -3,11 +3,10 @@ from __future__ import annotations
 
 import asyncio
 from asyncio import Future
-from asyncio.exceptions import CancelledError
 from typing import Any
 
 from pyweatherflowudp.client import EVENT_DEVICE_DISCOVERED, WeatherFlowListener
-from pyweatherflowudp.errors import AddressInUseError, ListenerError
+from pyweatherflowudp.errors import AddressInUseError, EndpointError, ListenerError
 
 from homeassistant import config_entries
 from homeassistant.core import callback
@@ -18,14 +17,13 @@ from .const import DOMAIN
 
 async def _async_can_discover_devices() -> bool:
     """Return if there are devices that can be discovered."""
-    future_event: Future[bool] = asyncio.get_running_loop().create_future()
+    future_event: Future[None] = asyncio.get_running_loop().create_future()
 
     @callback
     def _async_found(_):
         """Handle a discovered device - only need to do this once so."""
-
         if not future_event.done():
-            future_event.set_result(True)
+            future_event.set_result(None)
 
     async with WeatherFlowListener() as client, asyncio.timeout(10):
         try:
@@ -33,9 +31,6 @@ async def _async_can_discover_devices() -> bool:
             await future_event
         except asyncio.TimeoutError:
             return False
-        except CancelledError:
-            raise ListenerError
-
     return True
 
 
@@ -48,27 +43,30 @@ class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
         self, user_input: dict[str, Any] | None = None
     ) -> FlowResult:
         """Handle a flow initialized by the user."""
-
-        # Only allow a single instance of integration
+        # Only allow a single instance of integration since the listener
+        # will pick up all devices on the network and we don't want to
+        # create multiple entries.
         if self._async_current_entries():
             return self.async_abort(reason="single_instance_allowed")
 
-        # Get current discovered entries.
-        in_progress = self._async_in_progress()
-
-        if not (has_devices := in_progress):
-            errors = {}
+        errors: dict[str, str] = {}
+        if user_input is not None:
             try:
-                has_devices = await _async_can_discover_devices()  # type: ignore[assignment]
-            except AddressInUseError:
-                errors["base"] = "address_in_use"
-            except ListenerError:
-                errors["base"] = "cannot_connect"
-
-            if errors or (not has_devices and user_input is None):
-                return self.async_show_form(step_id="user", errors=errors)
-
+                has_devices = await _async_can_discover_devices()
+            except AddressInUseError as ex:
+                return self.async_abort(
+                    reason="address_in_use", description_placeholders={"error": ex}
+                )
+            except EndpointError as ex:
+                return self.async_abort(
+                    reason="endpoint_error", description_placeholders={"error": ex}
+                )
+            except ListenerError as ex:
+                return self.async_abort(
+                    reason="listener_error", description_placeholders={"error": ex}
+                )
             if not has_devices:
                 return self.async_abort(reason="no_devices_found")
-
             return self.async_create_entry(title="WeatherFlow", data={})
+
+        return self.async_show_form(step_id="user", errors=errors)
```

suggested simplifications (but I have no way to test so more work may be needed)

Reply:
If the exceptions don't have a useful message, drop the `description_placeholders` but we may want the message to go to the UI if it will help the user figure out what went wrong.  Ideally its translatable but its still better to get the raw message than nothing at all

Reply:
So I just pushed new code and came up with the following:

## No devices discovered

<img width="409" alt="image" src="https://github.com/home-assistant/core/assets/6491743/06d319fa-874a-49d7-ac3c-f8c9a630e691">

## Already configured
<img width="313" alt="image" src="https://github.com/home-assistant/core/assets/6491743/541e56b2-1cd2-42b7-965b-62fa205b646f">

## Address in use

<img width="541" alt="image" src="https://github.com/home-assistant/core/assets/6491743/c3c50ee3-d574-45a9-86df-643618e7df16">


What I'm unsure about is:

1) I configure the integration
2) Stop HA
3) Lock the UDP port via `nc -lu -p 50222`
4) Run Home assistant ...
5) I don't really get a good error message. I'm assuming I need an `async_abort` somewhere... because this isn't so helpful

<img width="598" alt="image" src="https://github.com/home-assistant/core/assets/6491743/2fc21ef2-bf1e-4653-84f1-49fc6a61d1bf">

I probalby need an error message which is the async_abort

Reply:
This was after my discord discussion with @joostlek ...

Reply:
Oh look it does have the error message:

<img width="186" alt="image" src="https://github.com/home-assistant/core/assets/6491743/bafe1469-8478-44f8-bcdb-088ee3eba432">

Reply:
But that is part of the `__init__.py` not of the config flow right

Reply:
Can we close this conversation out?

Reply:
It looks like `EndpointError` isn't handled?

Reply:
handled now! (assuming my tests pass)